cmake: Add support for file suffixes

nordicjm opened the pull request:
cmake: Add function for file suffix checking/appending
    
    Adds a function that can be used to check if a file suffix is
    supplied and, if so, will update a variable with these filenames


    cmake: Add support for file suffixes
    
    Adds support for configuration files to use file suffixes

Alternative to #65561

nordicjm committed:
cmake: Add function for file suffix checking/appending

Adds a function that can be used to check if a file suffix is
supplied and, if so, will update a variable with these filenames

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
cmake: Add support for file suffixes

Adds support for configuration files to use file suffixes

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented:
Thanks for raising this.

Before extending how conf files and friends are being picked up (the implementation), could we then have an issue describing the feature which is needed, so that we can see if it covers the needed use-cases, and / or discuss if such use cases can be handled with the snippets infrastructure.

Also it should be discussed if existing methods (like the current CONF_FILE_BUILD_TYPE) should be deprecated in favor of a new approach, but in order to be able to determine that we need to understand the use-cases.

The current CONF_FILE_BUILD_TYPE originates from here: https://github.com/zephyrproject-rtos/zephyr/issues/27934 with corresponding fix here: https://github.com/zephyrproject-rtos/zephyr/pull/28039.
This was later re-worked into the CONF_FILE_BUILD_TYPE variable which actually allows other parts of build system to test the value of such setting.

Is something similar considered for `FILE_SUFFIX` ?
If it is, then probably a better name than `FILE_SUFFIX` could be considered (i'm not having a better suggestion atm)

nordicjm commented:
> Also it should be discussed if existing methods (like the current CONF_FILE_BUILD_TYPE) should be deprecated in favor of a new approach, but in order to be able to determine that we need to understand the use-cases.

Commit was just updated to block using both, the existing functionality of `zephyr_file()` is reused with the build type set to the file suffix value. I do not envisage this as a replacement or to deprecate the existing functionality, this is designed with sysbuild in mind but even without sysbuild impacts all configuration files including dts overlay files which the build type does not

nordicjm commented:
> Before extending how conf files and friends are being picked up (the implementation), could we then have an issue describing the feature which is needed, so that we can see if it covers the needed use-cases, and / or discuss if such use cases can be handled with the snippets infrastructure.

Snippets cannot cover this use case

nordicjm committed:
sysbuild: Add support for FILE_SUFFIX

Adds supports for sysbuild loading project-specific Kconfiguration
fragments that are suffixed with the user-provided value

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand commented:
> > Before extending how conf files and friends are being picked up (the implementation), could we then have an issue describing the feature which is needed, so that we can see if it covers the needed use-cases, and / or discuss if such use cases can be handled with the snippets infrastructure.
> 
> Snippets cannot cover this use case

I understand the immediate use-case as to be able to have same suffix applied when using sysbuild.
Something which has not really been an issue before sysbuild because each image build would be completely independent and user would have had to do:
```
$ west build ... sampleA -DCONF_FILE=prj_foo.txt 
$ west build ... sampleB -DCONF_FILE=prj_foo.txt 
...
```
whereas with sysbuild that would be equivalent to:
```
$ west build ... sampleA -DCONF_FILE=prj_foo.txt -DsampleB_CONF_FILE=prj_foo.txt
```
and the wish then seems to be to avoid the `_foo` part and inherit that.

but what if snippets instead where having an image specific section, like `boards` has in snippets:
https://docs.zephyrproject.org/latest/build/snippets/design.html

Like:
```
# <app>/snippets/foo.yml
name: foo
images:
  mcuboot: # settings for image "mcuboot" goes here
    append:
      EXTRA_CONF_FILE: mcuboot.conf
  bar_image: # settings for image "bar_image" goes here
    append:
      EXTRA_CONF_FILE: bar.conf
```
could even be combined with a boards section under `images`.

I acknowledge that such work might be a bigger effort that the simple approach presented in this PR, but has it been considered if such extension to snippets would be a better option ?

nordicjm committed:
doc: Add documentation on FILE_SUFFIX build variable

Adds details to the application development page and sysbuild
page documenting the variable, how it works and how it should
be used

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.6: Add note on FILE_SUFFIX support

Adds a note with a link to the documentation pages for this
new feature

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm commented:
Snippets do not work for this, if we assume a project which has 3 sets of configurations each for 8 boards and a new image is added to that, this means depending upon combinations, at least 3 sets of snippets would need to be updated (24 at worst), and if you imagine 3 applications which use this feature but have different "suffixes" mean that a new snippet would be needed per configuration, per board, per application, so at best case you would be looking at 9 snippets or 72 at worst case. It simply does not work for this use case.

57300 commented on cmake/modules/configuration_files.cmake:
```diff
@@ -78,9 +78,15 @@ else()
   message(FATAL_ERROR "No prj.conf file was found in the ${APPLICATION_CONFIG_DIR} folder, "
                       "please read the Zephyr documentation on application development.")
 endif()
+zephyr_file_suffix(CONF_FILE)
```
This doesn't work consistently when I pass `-DCONF_FILE=<relative-path>`, as it depends on CMake's working directory. The same goes for other variables.

Sidenote: I caught this because sysbuild changes the working dir for each image, which in itself is confusing.

nordicjm replied:
Fixed

57300 commented on cmake/modules/configuration_files.cmake:
```diff
@@ -78,9 +78,15 @@ else()
 
 if(CONF_FILE_INCLUDE_FRAGMENTS)
-  zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR}/boards KCONF CONF_FILE BUILD ${CONF_FILE_BUILD_TYPE})
+  # FILE_SUFFIX will take priority over a prj_<X> build type if both are used
```
I think a better approach would be to disable the `prj_<X>` pattern up here:
https://github.com/zephyrproject-rtos/zephyr/blob/9102821fbe282d1b994ee21c0e3015c885de3d58/cmake/modules/configuration_files.cmake#L61

`if(NOT DEFINED FILE_SUFFIX AND ...`. I'm assuming we should deprecate that entire logic anyway, unless you see a reason to still support `prj_<X> -> prj_<X>_<suffix>`.

57300 commented on cmake/modules/configuration_files.cmake:
```diff
@@ -78,9 +78,15 @@ else()
-  zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR}/boards KCONF CONF_FILE BUILD ${CONF_FILE_BUILD_TYPE})
+  # FILE_SUFFIX will take priority over a prj_<X> build type if both are used
+  if(DEFINED FILE_SUFFIX)
+    zephyr_file(CONF_FILES ${APPLICATION_CONFIG_DIR}/boards KCONF CONF_FILE BUILD ${FILE_SUFFIX})
```
Passing `FILE_SUFFIX` as a build type looks hacky.

nordicjm replied:
I didn't want to add another variable to the function just to duplicate the code when it would be doing the same thing. Could rename it perhaps?

57300 commented on cmake/modules/configuration_files.cmake:
```diff
@@ -108,6 +118,7 @@ elseif(EXISTS          ${APPLICATION_CONFIG_DIR}/${BOARD}.overlay)
 elseif(EXISTS          ${APPLICATION_CONFIG_DIR}/app.overlay)
   set(DTC_OVERLAY_FILE ${APPLICATION_CONFIG_DIR}/app.overlay)
 endif()
+zephyr_file_suffix(DTC_OVERLAY_FILE)
```
Two issues with this:

1. Suffix may be applied twice, i.e., it's possible to find `boards/board_suffix_suffix.overlay`.
2. If I have an `app_suffix.overlay`, it can't be found unless `app.overlay` also exists.

57300 commented on cmake/modules/extensions.cmake:
```diff
@@ -2487,6 +2487,63 @@ function(zephyr_file_copy oldname newname)
+    # Loop for each list element
+    while(list_length GREATER 0)
+      list(POP_FRONT tmp_list tmp_element)
+      zephyr_file_suffix(tmp_element)
```
This doesn't need to be recursive. A single item should be handled as a list of length 1.

57300 commented on cmake/modules/root.cmake:
```diff
@@ -22,6 +22,7 @@ include_guard(GLOBAL)
 include(extensions)
 
 # Merge in variables from other sources (e.g. sysbuild)
+zephyr_get(FILE_SUFFIX SYSBUILD GLOBAL)
```
This doesn't belong here. How about `configuration_files.cmake`?

57300 commented on share/sysbuild/cmake/modules/sysbuild_extensions.cmake:
```diff
@@ -268,10 +268,13 @@ function(ExternalZephyrProject_Add)
+    # when user is not specifying custom fragments. Note: FILE_SUFFIX will take priority over
+    # a prj_<X> build type if both are used
+    if(DEFINED FILE_SUFFIX AND EXISTS "${sysbuild_image_conf_dir}/${ZBUILD_APPLICATION}_${FILE_SUFFIX}.conf")
+      set(sysbuild_image_conf_fragment ${sysbuild_image_conf_dir}/${ZBUILD_APPLICATION}_${FILE_SUFFIX}.conf)
```
By applying a `suffix` in both sysbuild and the image itself, you may end up using `image_suffix_suffix.conf`. Why is the sysbuild change needed?

nordicjm replied:
This is if someone has `<app>/sysbuild/mcuboot_<X>.conf` rather than `<app>/sysbuild/mcuboot/prj_<X>.conf` so is handled in sysbuild

nordicjm replied:
Fixes

57300 reviewed (changes requested):
Putting the snippets discussion aside, I think this works pretty well as a successor to `CONF_FILE_BUILD_TYPE`. The implementation needs work, but I'm not opposed to the idea.

I'm not fond of the name `FILE_SUFFIX` though. Could we come up with a more intuitive one?
* `FILE`: It would be nice to describe which file(s) this applies to, i.e., configuration files in Zephyr.
* `SUFFIX`: In many contexts, a "file suffix" is the same thing as a "file extension" (examples: `CMAKE_EXECUTABLE_SUFFIX`, Python's pathlib). But this suffix is meant to be inserted before the dot.

bencefr commented:
I would rather vote for a solution that fixes the original problem in the first place rather than introducing a new rule in the playbook that VS Code extension can hardly keep up with. As these rules are changing without our knowledge the feature we provide are already broken.

We need to guess from the cmake cache what the original parameters of the build were and match that to what the build system default variables would be. The extension must do this for all supported SDK versions out there, and we already see broken features due to changes to what the cache actually contains.

Please consider that majority of the users don't use command line tools anymore but rather an IDE, and that IDE must be supported.

nordicjm commented:
> I would rather vote for a solution that fixes the original problem in the first place rather than introducing a new rule in the playbook that VS Code extension can hardly keep up with. As these rules are changing without our knowledge the feature we provide are already broken.
> 
> We need to guess from the cmake cache what the original parameters of the build were and match that to what the build system default variables would be. The extension must do this for all supported SDK versions out there, and we already see broken features due to changes to what the cache actually contains.
> 
> Please consider that majority of the users don't use command line tools anymore but rather an IDE, and that IDE must be supported.

The idea was discussed with Trond and he raised no such issue. The suggestion for a future improvement *for IDE support* is to have a .yml file which lists what the possible options are. As for guessing the parameters, the IDE should not be guessing anything, if you are "guessing" parameters by looking at the cmake cache then your system is not right

nordicjm commented:
> > The idea was discussed with Trond and he raised no such issue. The suggestion for a future improvement _for IDE support_ is to have a .yml file which lists what the possible options are. As for guessing the parameters, the IDE should not be guessing anything, if you are "guessing" parameters by looking at the cmake cache then your system is not right
> 
> We have to guess because there's no other source of information. Anything you'd invent for the future doesn't help on the existing versions that we also have to support.

This seems like a problem with the design of the plugin. Big name IDEs all have a project or session file which is used to store configuration in, that includes visual studio, keil, qt, eclipse, netbrains intellij, etc. none of them "guess" the configuration

nordicjm commented:
@tejlmand @57300 please re-review

tejlmand commented on cmake/modules/extensions.cmake:
```diff
@@ -2487,6 +2487,72 @@ function(zephyr_file_copy oldname newname)
+      set(new_filename "${filename_beginning}_${FILE_SUFFIX}${filename_ending}")
+    endif()
+
+    # Check if the provided path is relative, if so, prepend the application configuration
+    # directory to it
+    cmake_path(IS_RELATIVE new_filename relative_path)
```
also, let's not re-do old errors.

We have had lengthy discussion as to `relative to what` means, and therefore we only allow relative paths for user defined settings.
All settings done within CMake, like a `set(file path)` should be done using abs path, for example `set(file ${CMAKE_CURRENT_SOURCE_DIR}/path)`. No relative paths in such cases.

Take a look here: https://github.com/zephyrproject-rtos/zephyr/blob/62809d758160baed837b6589ff8122cedc29951d/cmake/modules/extensions.cmake#L2411-L2412

https://github.com/zephyrproject-rtos/zephyr/blob/62809d758160baed837b6589ff8122cedc29951d/cmake/modules/extensions.cmake#L2482-L2485

nordicjm replied:
This is needed because this runs before the path has been transformed by the other functions, so it need to check if a filename with a suffix exists before or it will miss. It's been months since I looked but it was along the lines of if you have a `prj_<suffix>.conf` in a sysbuild configuration folder but no `prj.conf` then this is needed because without it, sysbuild will look for a `prj.conf` file which fails then revert back to the normal application's configuration directory which then won't find the out-of-tree `prj_<suffix>.conf` file

nordicjm replied:
Here we go, creating `prj2.conf` in `samples/subsys/mgmt/mcumgr/smp_svr` then running using `cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -DAPP_DIR=.. -DFILE_SUFFIX=test -Dsmp_svr_CONF_FILE="prj2.conf" /tmp/aa/zephyr/share/sysbuild/`:
```
   *****************************
   * Running CMake for smp_svr *
   *****************************
...
CMake Warning at /tmp/aa/zephyr/cmake/modules/extensions.cmake:2629 (message):
  Check: prj2.conf
...
CMake Warning at /tmp/aa/zephyr/cmake/modules/extensions.cmake:2644 (message):
  RELATIVE SKIP
...
Loaded configuration '/tmp/aa/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
Merged configuration '/tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/prj2.conf'
Merged configuration '/tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/_AA/smp_svr/zephyr/.config.sysbuild'
```

tejlmand replied:
> `-DFILE_SUFFIX=test -Dsmp_svr_CONF_FILE="prj2.conf"`

This is a mess imo, if an explicit conf file has been set with `CONF_FILE` by the user, then we should not start messing with that.
In this specific case, the user should use `-Dsmp_svr_CONF_FILE="prj2_test.conf"`

`FILE_SUFFIX` should be for the cases where the system is looking up a default set of files because none is given by user, meaning for example system sould be using `prj.conf`, `boards/<board>.overlay`, etc., but because user has set `FILE_SUFFIX=foo`, then `prj_foo.conf`, `boards/<board>_foo.overlay`, etc is looked up in addition.

nordicjm replied:
> meaning for example system sould be using prj.conf, boards/<board>.overlay, etc., but because user has set FILE_SUFFIX=foo, then prj_foo.conf, boards/<board>_foo.overlay, etc is looked up in addition

Do not agree, this is a worthless feature if it can only add files **in addition**, the point would be to allow for different configuration regardless without imposing anything on the user, with this rule it means they must construct their project in a way that essentially prj.conf is useless and a FILE_SUFFIX must be provided for every build

nordicjm replied:
> an example, a user wants to build with debug confs, except for foo image which should still use prj.conf.
> Such user should be able to do:

Well no a user should target it properly using `cmake -GNinja ... -Dabc_FILE_SUFFIX=debug <blah>` or not have a `prj_debug.conf` for `foo`

tejlmand replied:
> Well no a user should target it properly using `cmake -GNinja ... -Dabc_FILE_SUFFIX=debug <blah>` or not have a `prj_debug.conf` for `foo`

There can be many reason why you want to do a given build, let it be `debug`, `release`, `fast`, etc, and have everything constant but point a single image to be build different.
Like:
```
west build --sysbuild -DSUFFIX=foo ....
```
Keep everything the same, but now adjust only for bar image:
```
west build --sysbuild -DSUFFIX=foo -Dbar_<setting>=<test-file> ....
```

The more areas that are making use of file suffix, the higher the chance that a user needs just to explicitly point to a given file grows.

Therefore, if a user explicit points to a given file, then we should honor that.

> with this rule it means they must construct their project in a way that essentially prj.conf is useless and a FILE_SUFFIX must be provided for every build.

why does a user need to construct their project in a more special way than before ?
System still fall-back to default files, for example `prj.conf` if `prj_<suffix>.conf` is not found.

Only the case where user explicit specify a given file using `-D<setting>=<use-this-file>`, then we should honor that file.

nordicjm replied:
Because if you want to build normally, you do that, if you want to build in debug mode and files are set in e.g. cmake or a module for what to include, you then need to fork those modules and add code for handling it. Let's say someone creates a module `logger` and has `release` and `debug` build for it whereby `release` disables it and `debug` enables it and the module just adds the base .conf file, with my way `-DFILE_SUFFIX=release` and `-DFILE_SUFFIX=debug` work out of the box, with your way it cannot work at all

tejlmand reviewed (changes requested):
Reviewed implementation.

The intention of the feature is 👍 
However i'm really not fond of the name of the feature / CMake variable, `FILE_SUFFIX`.

Is that really the best we can come up with ?
Unfortunately I still haven't found a good name either.

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2487,6 +2487,72 @@ function(zephyr_file_copy oldname newname)
+      string(SUBSTRING "${${filename}}" 0 ${full_stop_position} filename_beginning)
+      string(SUBSTRING "${${filename}}" ${full_stop_position} -1 filename_ending)
+      set(new_filename "${filename_beginning}_${FILE_SUFFIX}${filename_ending}")
+    endif()
```
Fixed

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2487,6 +2487,72 @@ function(zephyr_file_copy oldname newname)
+# Returns an updated variable/list of absolute path(s)
+#
+function(zephyr_file_suffix filename)
+  if(NOT DEFINED FILE_SUFFIX OR NOT DEFINED ${filename})
```
Fixed

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2487,6 +2487,72 @@ function(zephyr_file_copy oldname newname)
+# `_<FILE_SUFFIX>` extension to them and if so, updates the supplied variable/list with the new
+# path/paths.
+#
+# <filename>: Variable or list of absolute path filename(s) which should be checked and updated if
```
Fixed

nordicjm commented on cmake/modules/extensions.cmake:
```diff
@@ -2487,6 +2487,72 @@ function(zephyr_file_copy oldname newname)
+
+    # Loop for each list element
+    while(list_length GREATER 0)
+      list(POP_FRONT tmp_list tmp_element)
```
Fixed

nordicjm commented on cmake/modules/configuration_files.cmake:
```diff
@@ -27,6 +27,9 @@ include_guard(GLOBAL)
 include(extensions)
 
+# Merge in variables from other sources (e.g. sysbuild)
+zephyr_get(FILE_SUFFIX SYSBUILD GLOBAL)
```
I do not think it should, for the reason that `zephyr_file_suffix` is completely reusable from modules and out of tree code and has no dependency on files being dts or Kconfig fragment files, it can be used to check for any file type which might be used for any purpose

tejlmand replied:
if you want a general function available in addition then that is acceptable, but remove the relative path adjustment and have the caller of the function adjust to abs path before calling this function.

zephyr file can then do the path adjustment, and afterwards use the shared function. 

Else there is a high risk that we end back in the rabbit whole where we have been stuck before.

There is no reason why CMake code should using relative paths, there are plenty of CMake variables to use for identifying where a file should reside, like `${CMAKE_CURRENT_SOURCE_DIR}/<file>`, `${APPLICATION_CONFIG_DIR}/<file>`, etc.

nordicjm replied:
Have updated

nordicjm committed:
cmake: modules: extensions: Deprecate build type

Adds a deprecation notice if a file is found which has the build
type informing the user of the replacement

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: release: 3.6: Add deprecation note for build type

Adds a note that build type has been deprecated in favour of
file suffix

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

tejlmand committed:
cmake: extend zephyr_file(CONF_FILES ...) to take NAMES as argument

Extend zephyr_file(CONF_FILES ...) to take a NAMES list of file names
to find instead of creating file names based on board and revision.

This allows to unify lookup prj.conf and <board>/app.overlay for
application, as well as pave the way for future enhancements to
configuration file handling.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
